### ISSUE TITLE
selflabel script is not performing well compare to scan script

### ISSUE THREAD

**Issue by gauravjain10** [wvangansbeke/Unsupervised-Classification#81]
scan script was giving below classification report
printing confusion matrix:

[[389 212]
 [222 177]]
              precision    recall  f1-score   support

     class 0       0.65      0.64      0.64       611
     class 1       0.44      0.46      0.45       389

    accuracy                           0.57      1000
   macro avg       0.55      0.55      0.55      1000
weighted avg       0.57      0.57      0.57      1000

{'ACC': 0.566, 'ARI': 0.014542700083472906, 'NMI': 0.0061884390657268576, 'ACC Top-2': 1.0, 'hungarian_match': [(0, 1), (1, 0)]}
#####################################################################################################################################
self label script was giving below classification report
printing confusion matrix:

[[601   0]
 [399   0]]

              precision    recall  f1-score   support

     class 0       1.00      0.60      0.75      1000
     class 1       0.00      0.00      0.00         0

    accuracy                           0.60      1000
   macro avg       0.50      0.30      0.38      1000
weighted avg       1.00      0.60      0.75      1000

{'ACC': 0.601, 'ARI': 0.0, 'NMI': 9.90380220639017e-16, 'ACC Top-2': 1.0, 'hungarian_match': [(0, 0), (1, 1)]}

######################################################################################
scan config file 
# setup
setup: scan

# Loss
criterion: scan
criterion_kwargs:
   entropy_weight: 5.0

# Weight update
update_cluster_head_only: True # Train only linear layer during SCAN
num_heads: 1 # Use single heads

# Model
backbone: resnet18

# Dataset
train_db_name: custom
val_db_name: custom
num_classes: 2
num_neighbors: 20

# Transformations
augmentation_strategy: simclr
augmentation_kwargs:
   random_resized_crop:
      size: 96
      scale: [0.2, 1.0]
   color_jitter_random_apply:
      p: 0.8
   color_jitter:
      brightness: 0.4
      contrast: 0.4
      saturation: 0.4
      hue: 0.1
   random_grayscale:
      p: 0.2
   normalize:
      mean: [0.0310, 0.0139, 0.9400] 
      std: [0.1164,0.0680,  0.1152]

transformation_kwargs:
   crop_size: 96
   normalize:
      mean: [0.0310, 0.0139, 0.9400] 
      std: [0.1164,0.0680,  0.1152]

# Hyperparameters
optimizer: adam
optimizer_kwargs:
   lr: 0.0001
   weight_decay: 0.0001
epochs: 50
batch_size: 10
num_workers: 8

# Scheduler
scheduler: constant

######################################################################################
selflabel config  file
# setup
setup: selflabel

# Threshold
confidence_threshold: 0.5

# ema
use_ema: False

# Loss
criterion: confidence-cross-entropy
criterion_kwargs:
   apply_class_balancing: True

# Model
backbone: resnet18
num_heads: 1

# Dataset
train_db_name: custom 
val_db_name: custom
num_classes: 2

# Transformations

augmentation_strategy: ours 
augmentation_kwargs:
   crop_size: 96
   normalize:

      mean: [0.0310, 0.0139, 0.9400] 
      std: [0.1164,0.0680,  0.1152]

   num_strong_augs: 4
   cutout_kwargs:
     n_holes: 1
     length: 32
     random: True

transformation_kwargs:
   crop_size: 96
   normalize:
      mean: [0.0310, 0.0139, 0.9400] 
      std: [0.1164,0.0680,  0.1152]

optimizer: adam
optimizer_kwargs:
   lr: 0.0001
   weight_decay: 0.0001
epochs: 50
batch_size: 10
num_workers: 8

# Scheduler
scheduler: constant
#################################################
I don't know why self label script is performing poorly
Total images i have is 1000 images, class 0 counts are 601 and class 1 counts are 399.
I have tried different confidence threshold like 0.99,0.9,0.8,0.7,0.6 etc but all of them giving allmask zero error then i even tried to increase batch size  like 10,50,100,500,and even checked for 1000 still it is predicting everything is class 0.
i have already checked different learning rate.
so can you tell me what else i can do

**Comment by wvangansbeke** [comment]
Hi @gauravjain10,

This makes sense to me. The results after SCAN are too low (55%) to obtain improvements with self-labeling. Keep in mind that in order to make self-labeling working well, you need a decent amount of confident samples per cluster (i.e. prototypes).

So you should first improve the results with SCAN. These results depend on the pretext task. It seems that the pretext task is unable to find the desired image similarities in your data. I don't know the details here (e.g., What type of dataset are you using?, Are the augmentations appropriate? etc.). These are extremely important factors. Furthermore, SSL frameworks like SimCLR or MoCo work particularly well if you have large amounts of data available with lots of variance/classes. 

However, your experiment might still work.
A solution might be to use a different and stronger set of data augmentations for the pretext task and/or a larger weight for the uniformity loss in SCAN. I would also try on more data, since I have some doubts about this experiment. If this doesn't work, you should test `SimCLR + KMeans` for comparison. I still need to upload that particular script, but it should be fairly easy to run KMeans in Python.

**Comment by gauravjain10** [comment]
Hi @wvangansbeke ,

Yeah right now images are similar to solar radio images, below is the given link. 
 http://soleil.i4ds.ch/solarradio/callistoQuicklooks/?date=20210101
Generally size of original images are 1200x800 and i am resizing them into 96 during data loader.
As you said i should try to improve pretext task clearly current configs is not able to get confident samples.
Although when images were nearly balanced like 475 images belongs to one class and 525 images belong to another class,there i was actually getting 100perc accuracy with same hyperparams given above.
SimCLR + KMeans That script would be good if you upload :)

Anyway thank you so much for quick response.
